File: src/main/java/io/github/divyeshmangla/timetable/model/TimeSlot.java
package io.github.divyeshmangla.timetable.model;

import java.time.LocalTime;
import java.time.format.DateTimeFormatter;

public enum TimeSlot {
    T1("08:00", "08:50"),
    T2("08:50", "09:40"),
    T3("09:40", "10:30"),
    T4("10:30", "11:20"),
    T5("11:20", "12:10"),
    T6("12:10", "13:00"),
    T7("13:00", "13:50"),
    T8("13:50", "14:40"),
    T9("14:40", "15:30"),
    T10("15:30", "16:20"),
    T11("16:20", "17:10"),
    ;

    private static final DateTimeFormatter FMT = DateTimeFormatter.ofPattern("HH:mm");

    public final LocalTime start;
    public final LocalTime end;

    TimeSlot(String start, String end) {
        this.start = LocalTime.parse(start);
        this.end = LocalTime.parse(end);
    }

    public static TimeSlot fromNumber(int number) {
        if (number < 1 || number > values().length) {
            throw new IllegalArgumentException("Invalid time slot number: " + number);
        }
        return values()[number - 1];
    }

    public String label() {
        return start.format(FMT) + "–" + end.format(FMT);
    }
}

File: internal/model/time_slot.go
package model

import (
	"fmt"
	"time"
)

type TimeSlot int

const (
	T1 TimeSlot = iota
	T2
	T3
	T4
	T5
	T6
	T7
	T8
	T9
	T10
	T11
)

var timeSlotTimes = []struct {
	start string
	end   string
}{
	{"08:00", "08:50"},
	{"08:50", "09:40"},
	{"09:40", "10:30"},
	{"10:30", "11:20"},
	{"11:20", "12:10"},
	{"12:10", "13:00"},
	{"13:00", "13:50"},
	{"13:50", "14:40"},
	{"14:40", "15:30"},
	{"15:30", "16:20"},
	{"16:20", "17:10"},
}

const timeFormat = "15:04"

func (ts TimeSlot) Start() time.Time {
	if ts < 0 || int(ts) >= len(timeSlotTimes) {
		return time.Time{}
	}
	t, _ := time.Parse(timeFormat, timeSlotTimes[ts].start)
	return t
}

func (ts TimeSlot) End() time.Time {
	if ts < 0 || int(ts) >= len(timeSlotTimes) {
		return time.Time{}
	}
	t, _ := time.Parse(timeFormat, timeSlotTimes[ts].end)
	return t
}

func FromNumber(number int) (TimeSlot, error) {
	if number < 1 || number > len(timeSlotTimes) {
		return 0, fmt.Errorf("invalid time slot number: %d", number)
	}

	return TimeSlot(number - 1), nil
}

func (ts TimeSlot) Label() string {
	if ts < 0 || int(ts) >= len(timeSlotTimes) {
		return ""
	}

	return fmt.Sprintf("%s–%s", timeSlotTimes[ts].start, timeSlotTimes[ts].end)
}
